377: Allow to translations to be imported with status "waiting"

akirk opened the pull request:
This adds a new dropdown to the import translations form:
<img width="332" alt="glotpress-import-waiting-current" src="https://cloud.githubusercontent.com/assets/203408/14179555/cbe7c3ae-f75f-11e5-98d7-15dcbe485135.png">

It also introduces two new permissions:
- `import`
- `import-waiting`

The latter permission will have the dropdown with just one entry, "waiting". If neither permission is set then importing is not possible and it will redirect away from the form.

Another bit changed: new translations entries will only be created if the translation doesn't yet exist in that current status (so for example if the entry already exists as "waiting" then it isn't duplicated, while this is the case without that part of the PR).

Resolves #377

akirk committed:
Closes #377: import translations as waiting

toolstack commented on gp-includes/routes/translation.php:
```diff
@@ -28,8 +28,15 @@ public function import_translations_get( $project_path, $locale_slug, $translati
+				$this->redirect_with_error( __( 'You are not allowed to do that!', 'glotpress' ), $url );
+				return;
+			}
 		}
```
I would think a permission called `import` would define if I can import at all, so if it is false it should redirect with error no matter what.

This entire logic structure could also be simplified by using `cannot_and_redirect()` for both checks and combining them in to a single if with an `||` operator.

akirk replied:
My thinking is:
- `import`: can import as current and waiting
- `import-waiting`: can only import as waiting

toolstack replied:
By default, you should assume the user has no permissions and then add them to the user by granting them.  The above permissions, for example, don't allow you to grant someone only rights to import current (aka the current behaviour).

toolstack commented on gp-templates/project-import.php:
```diff
@@ -24,9 +24,18 @@
+		$status_options[ 'current' ] = __( 'Current', 'glotpress' );
+	}
+	$status_dropdown = gp_select( 'status', $status_options, 'waiting' );
 ?>
```
If there is only one option, it shouldn't be displayed as a dropdown (probably not at all).

toolstack commented on gp-includes/routes/translation.php:
```diff
@@ -54,10 +61,37 @@ public function import_translations_post( $project_path, $locale_slug, $translat
+
+			return $status;
+		});
+
```
Instead of adding the filter, why not just pass in the desired status as part of the import call on line 113 below, like:

`
$translations_added = $translation_set->import( $translations, $status );
`

toolstack commented on gp-includes/things/translation-set.php:
```diff
@@ -137,12 +137,15 @@ public function import( $translations ) {
+
 		$current_translations_list = GP::$translation->for_translation( $this->project, $this, 'no-limit', array('status' => 'current', 'translated' => 'yes') );
-		$current_translations = new Translations();
+		$existing_translations['current'] = new Translations();
```
Is there a reason to rename the variable and make it an array?

akirk replied:
None other than unifying all existing translations, grouped by status, in a single array.

akirk replied:
Also, it simplifies the logic a little bit further down because then I can do:

```
$translated = $existing_translations[ $entry->status ]->translate_entry( $entry );
```

akirk commented:
@toolstack thanks for your review, I have addressed all of the items except `$existing_translations['current']`

ocean90 commented on gp-templates/project-import.php:
```diff
@@ -24,9 +24,17 @@
-	<dt><label	for="format"><?php _e( 'Format:', 'glotpress' ); ?></label></dt>
+	<dt><label for="format"><?php _e( 'Format:', 'glotpress' ); ?></label></dt>
 	<dd><?php echo $format_dropdown; ?></dd>
+	<?php if ( $can_import && ! $can_only_import_waiting ): ?>
```
> Notice: Undefined variable: can_import in /gp-templates/project-import.php on line 31

The template is also used to import originals. In this case we don't have to select a status and should instead hide it completely.

ocean90 commented on gp-includes/routes/translation.php:
```diff
@@ -28,7 +28,11 @@ public function import_translations_get( $project_path, $locale_slug, $translati
+		$can_only_import_waiting = ! $can_import && $this->can( 'import-waiting', 'translation-set', $translation_set->id );
+
+		if ( ! $can_import && ! $can_only_import_waiting ) {
+			$this->redirect_with_error( __( 'You are not allowed to do that!', 'glotpress' ), $url );
```
`$url` doesn't exist.

ocean90 commented on gp-includes/routes/translation.php:
```diff
@@ -54,7 +58,19 @@ public function import_translations_post( $project_path, $locale_slug, $translat
+		$can_only_import_waiting = ! $can_import && $this->can( 'import-waiting', 'translation-set', $translation_set->id );
+
+		if ( ! $can_import && ! $can_import_waiting ) {
+			$this->redirect_with_error( __( 'You are not allowed to do that!', 'glotpress' ), $url );
```
`$url` doesn't exist.

ocean90 commented on gp-includes/routes/translation.php:
```diff
@@ -28,7 +28,11 @@ public function import_translations_get( $project_path, $locale_slug, $translati
+		$can_import = $this->can( 'import', 'translation-set', $translation_set->id );
+		$can_only_import_waiting = ! $can_import && $this->can( 'import-waiting', 'translation-set', $translation_set->id );
+
+		if ( ! $can_import && ! $can_only_import_waiting ) {
```
The same logic should be added to https://github.com/GlotPress/GlotPress-WP/blob/873ff71b83b0d7abf7addda12b7ee8b14dcef5f5/gp-templates/translations.php#L237-L239 to avoid displaying a link which a user can't access.

akirk replied:
Done in https://github.com/GlotPress/GlotPress-WP/pull/378/files#diff-d32355c0a0d15e6d80d430ba54ea3362R211 and https://github.com/GlotPress/GlotPress-WP/pull/378/files#diff-c4239e7606696e35a13d1d0811859f71R237

akirk commented:
As an attempt to address conerns in #377 I removed the new permissions and added two new filters:
- `gp_can_import`
- `gp_import_with_status`

I also removed the status dropdown.

This still introduces the flexibility needed on translate.wordpress.org: the status can be set to `waiting` for users who should only have what was `import-waiting` before, and also users without `approve` permission can be given the right to import by modifying the value of `$can_approve`.

akirk commented:
@toolstack is this heading in the right direction or are you still having reservations?

toolstack commented:
@akirk you have a typo in the route, `apply_filter` should be `apply_filters`.

I think the rights model was the right way to go, just with positive permissions instead of negative ones.

However filters aren't bad either, though I wouldn't bother with `gp_can_import`, it's just a duplication of the `approve` permission.

Likewise `gp_import_with_status` would make more sense as `gp_import_status`.

akirk commented:
Ok, I have reverted back to using permissions:
- `approve` gives you the ability to import current and waiting
- `import-waiting` gives you the ability to import as waiting

So this retains the ability to import with the `approve` permission (with gaining the ability to import as waiting).

And `import-waiting` allows to give a user just that permission.

ocean90 commented:
This looks good so far. @akirk can you please fix the code styling issues? Some unit tests would be nice.

akirk commented:
I rebased and squashed to integrate #426, now tests pass fine.

toolstack commented on gp-includes/routes/translation.php:
```diff
@@ -28,7 +28,11 @@ public function import_translations_get( $project_path, $locale_slug, $translati
 
-		if ( $this->cannot_and_redirect( 'approve', 'translation-set', $translation_set->id ) ) {
+		$can_import_current = $this->can( 'approve', 'translation-set', $translation_set->id );
+		$can_import_waiting = $can_approve || $this->can( 'import-waiting', 'translation-set', $translation_set->id );
```
`$can_approve` hasn't been set yet.

toolstack commented on gp-templates/project-import.php:
```diff
@@ -24,9 +28,28 @@
+			<input type="hidden" name="status" value="<?php echo esc_attr( reset( array_keys( $status_options ) ) ); ?>" />
+			<?php echo esc_html( reset( array_values( $status_options ) ) ); ?>
+		<?php elseif ( count( $status_options ) > 1 ) : ?>
+			<?php echo gp_select( 'status', $status_options, 'waiting' ); ?>
```
Should the default value be 'current' instead of 'waiting' to match the existing behaviour?

akirk replied:
I agree, we shouldn't modify the existing behavior.

toolstack commented on gp-includes/routes/translation.php:
```diff
@@ -54,7 +58,26 @@ public function import_translations_post( $project_path, $locale_slug, $translat
 
-		if ( $this->cannot_and_redirect( 'approve', 'translation-set', $translation_set->id ) ) {
+		$can_import_current = $this->can( 'approve', 'translation-set', $translation_set->id );
+		$can_import_waiting = $can_approve || $this->can( 'import-waiting', 'translation-set', $translation_set->id );
```
$can_approve hasn't been set yet.

akirk committed:
Incorporate review feedback

akirk commented:
@toolstack thanks! I have fixed the two typos and changed the default dropdown value to `current`.
